.circleci/config.yml

```diff
@@ -637,16 +641,8 @@ jobs:
         type: string
     steps:
       - checkout
-      - unless:
-          condition:
-            or:
-              - equal: [ master, << pipeline.git.branch >> ]
-              - matches:
-                  pattern: "^release.*$"
-                  value: << pipeline.git.branch >>
-          steps:
-            - build-tools/merge-with-parent:
-                parent: master
+      - diff-and-merge-with-master:
```

Comment:
Removing this step would push the image with the commit hash of the tip of the feature branch

Reply:
how do you feel about removing this merge to master on the CI side? i feel like it's bifurcating the flow and making it more complex.

what i did just now was add a requirement that branches be up to date with master before being merged on Github. that probably solves this problem but probably at the layer we want to prevent conflicts

Reply:
Let's discuss at a jam session. Despite the branch policy, I think there's still value with merge-with-parent - incl a quicker turnaround to discovering issues (since rebasing with master is generally a last step). Last time the idea of a branch policy forcing people to be up to date with master was bounced around I remember there was push back from others, but maybe that's no longer the case.

I do agree though that the docker tagging here is adding a bit more complexity than I anticipated. I think it could be resolved if I pushed the commit after merging with parent back to the repo under a PR prefixed branch or something, but if it's not worth it then we can just tear this out

Reply:
RE: Bifurcation - it's worth noting there's already needed bifurcation for the difftool since we don't want to diff master against master etc.

Reply:
i mean we've observed that doing this step on CI effectively gave us no benefit for the use case we added it for, conflicting alembic versions made it to master anyway. plus two people spent several hours last week stumped by the discrepancy between the commit hash and the docker tag. At the very least we should make those tags consistent. Plus it only runs against master at time of commit, so by the time merge rolls around it's likely outdated anyway

I think we should rethink this approach

Reply:
from a process perspective, the only way we can guarantee master is green is either force PR's to pull master before merge. we can either do that by hand or have a merge queue

Reply:
the third option is maybe we don't see this often enough to warrant adding this level of process or tooling and instead we automate stage deploys from master every day and use health checks to figure out if something is broken

Reply:
> i mean we've observed that doing this step on CI effectively gave us no benefit 

I disagree that there was _no_ benefit, but I'll admit that it doesn't fully prevent the migration conflict scenario (which was known going in). When the conflict happened, once the fix went in, nobody had to go to their terminal, check out their feature branch(es) (if they had started working on something else), rebase off of master, and push - they could simply re-run CI and get the fix, which could even be done on their phone away from their laptop. It's not the ideal prevention scenario but I would argue still better than what we had prior.

> At the very least we should make those tags consistent.

Yes. I 100% agree. @csjiang has a great doc on this in Notion. My suggested solutions are either to tag the docker image with the commit hash of the tip of the feature branch (which introduces a discrepancy of what is actually in the tag, not good) or have CI push the resulting merge commit back to the origin so it can be found (which I can timebox and investigate). Also, this may be a hot take, but I also don't think we should be putting docker images from feature branches that haven't merged yet onto staging/prod. This change in this PR allows for hotfix branches to not merge with parent so that we can still quickly roll out an image for an unmerged branch (but hopefully that's rare) and have it have consistent tags.

> Plus it only runs against master at time of commit, so by the time merge rolls around it's likely outdated anyway

In my previous work, we'd make the results of the CI reset after a set amount of time (I think it was 24 hours) which helps mitigate this a bit. Admittedly it's less strict than the branch policy. A perfect blend of the branch policy and this alternative would be a policy that requires the CI to have run on latest master, but that's unfortunately not easily done.

> we can either do that by hand or have a merge queue

Yeah. I tried to call that out when I pitched merge-with-parent initially - having CI merge with parent is only one piece of the puzzle:
https://about.gitlab.com/blog/2019/09/11/how-to-avoid-broken-master-with-pipelines-for-merge-requests/

> the third option is maybe we don't see this often enough to warrant adding this level of process or tooling

I agree that maybe this is a bit too early for us to see the benefits of it at scale. I think right now there's less chance of conflicts and maybe less payoff with having an automated system prevent/remedy them when they happen since our team is small, so I'm ok if we want to tear this out. That said, there are still benefits in my opinion, and assuming we can fix the docker tag issue easily, what is the cost of keeping it?